### PR TITLE
Add Jdbc connection support

### LINKED ISSUE #44: Add support for JDBC Connections

### PR DESCRIPTION
**Pull request by wynnel**
Fixes #44

## Proposed Changes
* add jdbc connection

### CODE REVIEW

File: src/main/java/com/ca/apim/gateway/cagatewayconfig/tasks/zip/builder/BundleEntityBuilder.java [hunk 1]
```diff
@@ -13,6 +13,8 @@
 import org.w3c.dom.Document;
 import org.w3c.dom.Element;
 
+import java.util.Map;
```
Comment (kazvictor) [inline src/main/java/com/ca/apim/gateway/cagatewayconfig/tasks/zip/builder/BundleEntityBuilder.java]: Code Smell Code Smell: Remove this unused import 'java.util.Map'. ([squid:UselessImportCheck](https://sonarcloud.io/organizations/ca-api-gateway/rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=com.ca.apim.gateway%3Agateway-developer-plugin&pullRequest=53&issues=AWU6Y48VWWqqPQVjPd1w&open=AWU6Y48VWWqqPQVjPd1w)

[ISSUE_KEY]: AWU6Y48VWWqqPQVjPd1w

File: src/main/java/com/ca/apim/gateway/cagatewayconfig/tasks/zip/builder/JdbcConnectionEntityBuilder.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        extension.appendChild( buildPropertiesElement("l7:ConnectionProperties", Stream.concat(jdbc.getProperties()==null? Stream.empty():jdbc.getProperties().entrySet().stream(),Stream.of(
+            new AbstractMap.SimpleEntry<>("user", jdbc.getUsername()),
+            new AbstractMap.SimpleEntry<>("password", jdbc.getPassword()))).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)),document));
+
+        Entity entity =  new Entity("JDBC_CONNECTION", name, id, jdbcElement);
+        entity.setMappingProperty(Entity.MAPPING_PROPERTY_MAP_BY, "name");
+        return entity;
+    }
+
+    // todo refactor after commit with utils
```
Comment (kazvictor) [inline src/main/java/com/ca/apim/gateway/cagatewayconfig/tasks/zip/builder/JdbcConnectionEntityBuilder.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://sonarcloud.io/organizations/ca-api-gateway/rules#rule_key=squid%3AS1135))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=com.ca.apim.gateway%3Agateway-developer-plugin&pullRequest=53&issues=AWU6Y48aWWqqPQVjPd1y&open=AWU6Y48aWWqqPQVjPd1y)

[ISSUE_KEY]: AWU6Y48aWWqqPQVjPd1y

File: src/main/java/com/ca/apim/gateway/cagatewayconfig/tasks/zip/loader/JdbcConnectionLoader.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ * Copyright (c) 2018 CA. All rights reserved.
+ * This software may be modified and distributed under the terms
+ * of the MIT license.  See the LICENSE file for details.
+ */
+
+package com.ca.apim.gateway.cagatewayconfig.tasks.zip.loader;
+
+import com.ca.apim.gateway.cagatewayconfig.tasks.zip.beans.Bundle;
+import com.ca.apim.gateway.cagatewayconfig.tasks.zip.beans.JdbcConnection;
+import com.ca.apim.gateway.cagatewayconfig.tasks.zip.beans.Service;
```
Comment (kazvictor) [inline src/main/java/com/ca/apim/gateway/cagatewayconfig/tasks/zip/loader/JdbcConnectionLoader.java]: Code Smell Code Smell: Remove this unused import 'com.ca.apim.gateway.cagatewayconfig.tasks.zip.beans.Service'. ([squid:UselessImportCheck](https://sonarcloud.io/organizations/ca-api-gateway/rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=com.ca.apim.gateway%3Agateway-developer-plugin&pullRequest=53&issues=AWU6Y48iWWqqPQVjPd10&open=AWU6Y48iWWqqPQVjPd10)

[ISSUE_KEY]: AWU6Y48iWWqqPQVjPd10

File: src/main/java/com/ca/apim/gateway/cagatewayconfig/tasks/zip/builder/JdbcConnectionEntityBuilder.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+                new AbstractMap.SimpleEntry<>("minimumPoolSize", jdbc.getMinimumPoolSize()))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)), document));
+
+
+        Element extension  = document.createElement("l7:Extension");
+        extension.appendChild(DocumentTools.createElement(document, "l7:DriverClass", jdbc.getDriver()));
+        extension.appendChild(DocumentTools.createElement(document, "l7:JdbcUrl", jdbc.getUrl()));
+        jdbcElement.appendChild(extension);
+
+        // todo password as secrets?
```
Comment (kazvictor) [inline src/main/java/com/ca/apim/gateway/cagatewayconfig/tasks/zip/builder/JdbcConnectionEntityBuilder.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://sonarcloud.io/organizations/ca-api-gateway/rules#rule_key=squid%3AS1135))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=com.ca.apim.gateway%3Agateway-developer-plugin&pullRequest=53&issues=AWU6Y48aWWqqPQVjPd1x&open=AWU6Y48aWWqqPQVjPd1x)

[ISSUE_KEY]: AWU6Y48aWWqqPQVjPd1x

File: src/main/java/com/ca/apim/gateway/cagatewayconfig/tasks/zip/builder/JdbcConnectionEntityBuilder.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        this.idGenerator = idGenerator;
+    }
+
+    public List<Entity> build(Bundle bundle) {
+        return bundle.getJdbcconnections().entrySet().stream().map(jdbcConnectionEntry ->
+                buildEntity(bundle, jdbcConnectionEntry.getKey(), jdbcConnectionEntry.getValue())
+        ).collect(Collectors.toList());
+    }
+
+    private Entity buildEntity(Bundle bundle, String name, JdbcConnection jdbc) {
```
Comment (kazvictor) [inline src/main/java/com/ca/apim/gateway/cagatewayconfig/tasks/zip/builder/JdbcConnectionEntityBuilder.java]: Code Smell Code Smell: Remove this unused method parameter "bundle". ([squid:S1172](https://sonarcloud.io/organizations/ca-api-gateway/rules#rule_key=squid%3AS1172))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=com.ca.apim.gateway%3Agateway-developer-plugin&pullRequest=53&issues=AWU6Y48aWWqqPQVjPd1z&open=AWU6Y48aWWqqPQVjPd1z)

[ISSUE_KEY]: AWU6Y48aWWqqPQVjPd1z

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/tasks/zip/builder/JdbcConnectionEntityBuilder.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+import java.util.stream.Collectors;
+
+import static com.ca.apim.gateway.cagatewayconfig.util.gateway.BuilderUtils.buildAndAppendPropertiesElement;
+import static com.ca.apim.gateway.cagatewayconfig.util.gateway.BuilderUtils.buildPropertiesElement;
+import static com.ca.apim.gateway.cagatewayconfig.util.gateway.BundleElementNames.*;
+import static com.ca.apim.gateway.cagatewayconfig.util.xml.DocumentUtils.*;
+
+public class JdbcConnectionEntityBuilder implements EntityBuilder {
+
+    private static final String STORED_PASSWORD_REF_FORMAT = "${secpass.%s.plaintext}";
```
Comment (apigatewayopensource) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/tasks/zip/builder/JdbcConnectionEntityBuilder.java]: Vulnerability Vulnerability: 'PASSWORD' detected in this expression, review this potentially hard-coded credential. ([squid:S2068](https://sonarcloud.io/organizations/ca-api-gateway/rules#rule_key=squid%3AS2068))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=com.ca.apim.gateway%3Agateway-developer-plugin&pullRequest=53&issues=AWXznoBUOlZ5QoPzBphu&open=AWXznoBUOlZ5QoPzBphu)

[ISSUE_KEY]: AWXznoBUOlZ5QoPzBphu

File: doc/jdbc-connections.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
+# JDBC Connections Configuration
+This is a file containing the JDBC Connections that are available in the gateway.
```
Comment (yves-chan) [inline doc/jdbc-connections.md]: Not sure if we should put export documentation here, but we should mention that the jdbc will only be exported when it's referenced by a policy

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/tasks/zip/beans/Bundle.java [hunk 8]
```diff
@@ -105,6 +106,17 @@ public void putAllStoredPasswords(Map<String, StoredPassword> storedPasswords) {
         this.storedPasswords.putAll(storedPasswords);
     }
 
+    public Map<String, JdbcConnection> getJdbcConnections() {
+        return jdbcConnections;
+    }
+
+    public void putAllJdbcConnections(Map<String, JdbcConnection> jdbcConnections) {
+        if (jdbcConnections == null) {
```
Comment (yves-chan) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/tasks/zip/beans/Bundle.java]: if jdbc connections is null, can we still do putAll?

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/tasks/zip/loader/EntityLoaderRegistry.java [hunk 9]
```diff
@@ -29,6 +29,7 @@ public EntityLoaderRegistry(FileUtils fileUtils, JsonTools jsonTools, IdGenerato
         loadersCollection.add(new IdentityProviderLoader(jsonTools));
         loadersCollection.add(new ListenPortLoader(jsonTools));
         loadersCollection.add(new StoredPasswordsLoader(fileUtils));
+	loadersCollection.add(new JdbcConnectionLoader(jsonTools));
```
Comment (yves-chan) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/tasks/zip/loader/EntityLoaderRegistry.java]: indentation is off here

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/util/gateway/BuilderUtils.java [hunk 10]
```diff
@@ -76,10 +84,6 @@ static Element buildPropertiesElement(final Map<String, Object> properties, fina
             return emptyMap();
         }
 
-        if (!Objects.equals(propertiesElement.getNodeName(), PROPERTIES)) {
```
Comment (yves-chan) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/util/gateway/BuilderUtils.java]: We can probably keep this in here if we substitute for PROPERTIES for propertiesElementName

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/util/gateway/BundleElementNames.java [hunk 11]
```diff
@@ -109,6 +115,13 @@
     public static final String STORED_PASSWD = "l7:StoredPassword";
     public static final String PASSWORD = "l7:Password";
 
+    // JDBC Connections
+    public static final String JDBC_CONNECTION = "l7:JDBCConnection";
+    public static final String DRIVER_CLASS = "l7:DriverClass";
+    public static final String JDBC_URL = "l7:JdbcUrl";
+    public static final String CONNECTION_PROPERTIES = "l7:ConnectionProperties";
+
+
```
Comment (yves-chan) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/util/gateway/BundleElementNames.java]: extra line here

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/util/gateway/BundleElementNames.java [hunk 12]
```diff
@@ -19,11 +19,17 @@
...
+    public static final String PROPERTY_PASSWORD = "password";
+    public static final String PROPERTY_MIN_POOL_SIZE = "minimumPoolSize";
+    public static final String PROPERTY_MAX_POOL_SIZE = "maximumPoolSize";
+
     // Common Elements
     public static final String NAME = "l7:Name";
     public static final String PROPERTIES = "l7:Properties";
     public static final String PROPERTY = "l7:Property";
-    public static final String INT_VALUE = "l7:IntValue";
+    public static final String INT_VALUE = "l7:IntegerValue";
```
Comment (yves-chan) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/util/gateway/BundleElementNames.java]: Just want to confirm, is there l7:IntegerValue and l7:IntValue? if so should we keep both?

File: doc/jdbc-connections.md (hunk 7 above)
Comment (joao-borges) [reply doc/jdbc-connections.md]: I think we need to add documentation for the export plugin there, not here.

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/util/gateway/BundleElementNames.java (hunk 12 above)
Comment (joao-borges) [reply config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/util/gateway/BundleElementNames.java]: I checked the xsds for the bundles and the correct is IntegerValue.
see `gateway-management-7.1.xsd` on line 907

**Pull request merged**